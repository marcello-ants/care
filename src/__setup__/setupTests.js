import '@testing-library/jest-dom/extend-expect';

jest.mock('@care/analytics', () => {
  // Mock all but init function
  const actual = jest.requireActual('@care/analytics');
  return Object.keys(actual).reduce((acc, method) => {
    if (method === 'init') {
      acc[method] = actual.init;
    } else {
      acc[method] = jest.fn();
    }

    return acc;
  }, {});
});

jest.mock('@/components/hooks/useResizeObserver', () => ({
  __esModule: true,
  default: jest.fn().mockImplementation(() => ({
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn(),
  })),
}));

jest.mock('@/utilities/utagHelper', () => {
  return {
    TealiumUtagService: { view: () => {} },
  };
});

jest.mock('@/lib/clientLogger');

jest.mock('oidc-client');

jest.mock('next/config', () => () => {
  return {
    publicRuntimeConfig: {
      APP_NAME: 'APP_NAME',
      APP_VERSION: 'APP_VERSION',
      GRAPHQL_URL: 'GRAPHQL_URL',
      OIDC_STS_AUTHORITY: 'OIDC_STS_AUTHORITY',
      OIDC_CLIENT_ID: 'OIDC_CLIENT_ID',
      OIDC_RESPONSE_TYPE: 'OIDC_RESPONSE_TYPE',
      OIDC_CLIENT_SCOPE: 'OIDC_CLIENT_SCOPE',
      OIDC_CALLBACK_PATH: 'OIDC_CALLBACK_PATH',
      OIDC_LOGOUT_URL: 'OIDC_LOGOUT_URL',
      AMPLITUDE_API_KEY: 'AMPLITUDE_API_KEY',
      ANALYTICS_ENABLED: 'ANALYTICS_ENABLED',
      CZEN_GENERAL_LOGIN: 'CZEN_GENERAL_LOGIN',
      CZEN_GENERAL: 'CZEN_GENERAL',
      CZEN_API_KEY: 'CZEN_API_KEY',
      TEALIUM_ENV: 'TEALIUM_ENV',
      SENTRY_ENABLED: 'SENTRY_ENABLED',
      SENTRY_ENV: 'SENTRY_ENV',
      SENTRY_DSN: 'SENTRY_DSN',
      SENTRY_APP_RELEASE: 'SENTRY_APP_RELEASE',
    },
  };
});

beforeEach(() => {
  if (typeof window !== 'undefined') {
    window.localStorage?.clear();
    window.sessionStorage?.clear();
  }
});

// eslint-disable-next-line no-global-assign
global.fetch = jest.fn(() => Promise.resolve({}));

if (typeof document !== 'undefined') {
  const fbScript = document.createElement('script');
  fbScript.id = 'facebook-jssdk';
  document.body.appendChild(fbScript);
}
