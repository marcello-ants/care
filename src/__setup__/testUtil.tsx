/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { NextRouter } from 'next/router';
import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { RouterContext } from 'next/dist/shared/lib/router-context';
import DayJSUtils from '@date-io/dayjs';
import { theme } from '@care/material-ui-theme';
import { ThemeProvider } from '@material-ui/core';
import { FeatureFlags, FeatureFlagsProvider } from '@/components/FeatureFlagsContext';
import { AppStateProvider } from '@/components/AppState';
import { initialAppState } from '@/state';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import { AppState } from '@/types/app';

export interface PrepareTreeOptions {
  pathname?: string;
  mocks?: MockedResponse[];
  flags?: FeatureFlags;
  appState?: AppState;
  addTypeName?: boolean;
  basePath?: string;
}

const prepareTree = (options?: PrepareTreeOptions) => {
  const {
    pathname = '/path',
    mocks = [],
    flags = {},
    appState = initialAppState,
    addTypeName = false,
    basePath = '/',
  } = options || {};
  const routerMock = {
    push: jest.fn(),
    asPath: pathname,
    beforePopState: (cb: any) => {
      cb({ url: pathname });
    },
    basePath,
    pathname,
    query: {},
    back: jest.fn(),
  } as unknown as NextRouter;

  const renderFn = (children: React.ReactNode) => (
    <ThemeProvider theme={theme}>
      <RouterContext.Provider value={routerMock}>
        <MockedProvider mocks={mocks} addTypename={addTypeName}>
          <MuiPickersUtilsProvider utils={DayJSUtils}>
            <FeatureFlagsProvider flags={flags}>
              <AppStateProvider initialStateOverride={appState}>{children}</AppStateProvider>
            </FeatureFlagsProvider>
          </MuiPickersUtilsProvider>
        </MockedProvider>
      </RouterContext.Provider>
    </ThemeProvider>
  );

  return {
    renderFn,
    routerMock,
  };
};

export const preRenderHook = (options?: PrepareTreeOptions) => {
  const tree = prepareTree(options);
  const renderFn = <T,>(hook: () => T) =>
    renderHook(hook, {
      wrapper({ children }: { children: React.ReactNode }) {
        return tree.renderFn(children);
      },
    });

  return {
    ...tree,
    renderFn,
  };
};

export const preRenderPage = (options?: PrepareTreeOptions) => {
  const tree = prepareTree(options);
  const renderFn = (ComponentOrChildren: React.ComponentType | React.ReactNode) => {
    if (React.isValidElement(ComponentOrChildren)) {
      return render(tree.renderFn(ComponentOrChildren));
    }

    const Component = ComponentOrChildren as React.ComponentType;
    return render(tree.renderFn(<Component />));
  };

  return {
    ...tree,
    renderFn,
    renderTree: tree.renderFn,
  };
};

export const setupWindowLocation = () => {
  const originalLocation = window.location;
  const windowLocationMock = jest.fn();

  // @ts-ignore
  delete window.location;
  // @ts-ignore
  window.location = {
    assign: windowLocationMock,
  };

  return {
    mock: windowLocationMock,
    cleanUp() {
      window.location = originalLocation;
    },
  };
};

export const setupLocalStorage = () => {
  const originalLocalStorage: Storage = window.localStorage;
  const localStorageMock = jest.fn();
  Object.defineProperty(window, 'localStorage', localStorageMock);

  return {
    mock: localStorageMock,
    cleanUp() {
      Object.defineProperty(window, 'localStorage', originalLocalStorage);
    },
  };
};

export function mockFetch({ ok, status, data }: { ok: boolean; status: number; data: any }) {
  return () =>
    Promise.resolve({
      ok,
      status,
      json: () => Promise.resolve(data),
    } as Response);
}
