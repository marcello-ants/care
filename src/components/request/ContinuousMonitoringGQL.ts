import { gql } from '@apollo/client';

export const GET_CONTINUOUS_MONITORING_ENROLLMENT_STATUS = gql`
  query ContinuousMonitoringEnrollmentStatus($memberId: ID!) {
    continuousMonitoringEnrollmentStatus(memberId: $memberId) {
      ... on ContinuousMonitoringEnrollmentStatusSuccess {
        needsConsent
        consentMessagesList {
          consentMessageId
          message
        }
      }
    }
  }
`;

export const ENROLL_IN_CONTINUOUS_MONITORING = gql`
  mutation EnrollInContinuousMonitoring($input: EnrollInContinuousMonitoringInput!) {
    enrollInContinuousMonitoring(input: $input) {
      ... on EnrollInContinuousMonitoringSuccess {
        success
      }
    }
  }
`;
