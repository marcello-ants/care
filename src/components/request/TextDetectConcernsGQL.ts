import { gql } from '@apollo/client';

// apollo client breaks if this query uses union type inside fragments and silently drops detectedConcerns from results
const TEXT_DETECT_CONCERNS = gql`
  query textAnalyzerDetectConcerns(
    $bio: String!
    $headline: String!
    $textCreator: TextAnalyzerDetectConcernsCreator!
    $validationId: String!
  ) {
    bio: textAnalyzerDetectConcerns(
      textToAnalyze: { text: $bio }
      textCreator: $textCreator
      validationId: $validationId
    ) {
      ... on TextAnalyzerDetectConcernsSuccess {
        __typename
        detectedConcerns {
          endIndex
          startIndex
          text
          type
        }
        validationId
      }
      ... on TextAnalyzerDetectConcernsError {
        __typename
        message
      }
    }
    headline: textAnalyzerDetectConcerns(
      textToAnalyze: { text: $headline }
      textCreator: $textCreator
      validationId: $validationId
    ) {
      ... on TextAnalyzerDetectConcernsSuccess {
        __typename
        detectedConcerns {
          endIndex
          startIndex
          text
          type
        }
        validationId
      }
      ... on TextAnalyzerDetectConcernsError {
        __typename
        message
      }
    }
  }
`;

export { TEXT_DETECT_CONCERNS };
