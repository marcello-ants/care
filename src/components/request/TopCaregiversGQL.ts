import { gql } from '@apollo/client';

const GET_TOP_CAREGIVERS = gql`
  query getTopCaregivers(
    $zipcode: String!
    $serviceID: ServiceType!
    $numResults: Int!
    $hourlyRate: JobRateInput
    $hasCareCheck: Boolean
    $qualities: [SeniorCareProviderQuality!]
    $services: [SeniorCareServiceProvidedType!]
    $maxDistanceFromSeeker: DistanceInput
    $hasApprovedActivePhoto: Boolean
    $enableFuzzySearch: Boolean
    $daysSinceLastActive: Int
    $isActiveAccount: Boolean
    $minimumAvgReviewRating: Float
    $fullTime: Boolean
  ) {
    topCaregivers: getTopCaregivers(
      zipcode: $zipcode
      numResults: $numResults
      serviceType: $serviceID
      hourlyRate: $hourlyRate
      hasCareCheck: $hasCareCheck
      qualities: $qualities
      services: $services
      maxDistanceFromSeeker: $maxDistanceFromSeeker
      hasApprovedActivePhoto: $hasApprovedActivePhoto
      enableFuzzySearch: $enableFuzzySearch
      daysSinceLastActive: $daysSinceLastActive
      isActiveAccount: $isActiveAccount
      minimumAvgReviewRating: $minimumAvgReviewRating
      fullTime: $fullTime
    ) {
      caregiver {
        profileURL
        avgReviewRating
        numberOfReviews
        yearsOfExperience
        hasCareCheck
        responseTime
        member {
          address {
            city
            state
          }
          displayName
          firstName
          id
          imageURL
          legacyId
        }
        profiles {
          commonCaregiverProfile {
            id
          }
          childCareCaregiverProfile {
            bio {
              experienceSummary
            }
            qualities {
              certifiedNursingAssistant
              certifiedRegisterNurse
              certifiedTeacher
              childDevelopmentAssociate
              comfortableWithPets
              cprTrained
              crn
              doesNotSmoke
              doula
              earlyChildDevelopmentCoursework
              earlyChildhoodEducation
              firstAidTraining
              nafccCertified
              ownTransportation
              specialNeedsCare
              trustlineCertifiedCalifornia
            }
            supportedServices {
              carpooling
              craftAssistance
              errands
              groceryShopping
              laundryAssistance
              lightHousekeeping
              mealPreparation
              swimmingSupervision
              travel
            }
            payRange {
              hourlyRateFrom {
                amount
                currencyCode
              }
              hourlyRateTo {
                amount
                currencyCode
              }
            }
          }
        }
        seniorCareProviderProfile {
          bio
          services
          payRange {
            hourlyRateFrom {
              amount
              currencyCode
            }
            hourlyRateTo {
              amount
              currencyCode
            }
          }
          qualities
        }
        signUpDate
      }
      distanceFromRequestZip {
        unit
        value
      }
    }
  }
`;

export default GET_TOP_CAREGIVERS;
