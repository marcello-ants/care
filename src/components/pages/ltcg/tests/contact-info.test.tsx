import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Formik } from 'formik';
import { preRenderPage } from '@/__setup__/testUtil';
import { VALIDATE_MEMBER_EMAIL, VALIDATE_MEMBER_PASSWORD } from '@/components/request/GQL';
import ContactInfo from '../contact-info';

const emailValidationMock = {
  request: {
    query: VALIDATE_MEMBER_EMAIL,
    variables: {
      email: 'guest@care-test.com',
    },
  },
  result: {
    data: {
      validateMemberEmail: { errors: [] },
    },
  },
};

const passwordValidationMock = {
  request: {
    query: VALIDATE_MEMBER_PASSWORD,
    variables: {
      password: 'tav*awf6mfa9xmh@HXD',
    },
  },
  result: {
    data: {
      validateMemberPassword: { errors: [] },
    },
  },
};

let onSubmitMock: jest.Mock;
function renderComponent(props?: any) {
  onSubmitMock = jest.fn();
  const { renderFn } = preRenderPage({ mocks: [emailValidationMock, passwordValidationMock] });
  const utils = renderFn(
    <Formik
      initialValues={{
        email: '',
        password: '',
        phoneNumber: '',
      }}
      onSubmit={onSubmitMock}
      validateOnMount>
      <ContactInfo {...props} />
    </Formik>
  );
  return { ...utils };
}

describe('ContactInfo', () => {
  it('matches snapshot', () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it('starts with disabled button', async () => {
    renderComponent();
    const submitButton = screen.getByRole('button');
    await waitFor(() => expect(submitButton).toBeDisabled());
  });

  it('enables button with correct info', async () => {
    renderComponent();
    const emailInput = screen.getByLabelText('Email address');
    const phoneInput = screen.getByLabelText('Phone number');
    const passwordInput = screen.getByLabelText('Password');

    const submitButton = screen.getByRole('button');
    await waitFor(() => expect(submitButton).toBeDisabled());

    userEvent.type(emailInput, 'guest@care-test.com');
    userEvent.type(phoneInput, '3107645678');
    userEvent.type(passwordInput, 'tav*awf6mfa9xmh@HXD');

    await waitFor(() => expect(submitButton).toBeEnabled());
  });

  it('should call submit when the submit button is clicked', async () => {
    renderComponent();
    const submitButton = screen.getByRole('button');

    await waitFor(() => expect(submitButton).toBeDisabled());

    userEvent.type(screen.getByLabelText('Email address'), 'guest@care-test.com');
    userEvent.type(screen.getByLabelText('Phone number'), '3107645678');
    userEvent.type(screen.getByLabelText('Password'), 'tav*awf6mfa9xmh@HXD');

    await waitFor(() => expect(submitButton).toBeEnabled());
    userEvent.click(submitButton);

    await waitFor(() => expect(onSubmitMock).toHaveBeenCalled());
  });
});
