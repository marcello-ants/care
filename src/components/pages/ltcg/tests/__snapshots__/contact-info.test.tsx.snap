// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactInfo matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="MuiTypography-root makeStyles-header-4 MuiTypography-h2"
    >
      Share your contact info so we can help you through the process and set up your account.
    </h2>
    <p
      class="MuiTypography-root makeStyles-subheader-1 MuiTypography-body2"
    >
      Our experts will be in touch quickly to help identify your needs.
    </p>
    <div
      class="makeStyles-inputContainer-2"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-inline-5"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="email"
          id="email-label"
        >
          Email address
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="email"
            name="email"
            type="email"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="makeStyles-inputContainer-2"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-inline-5"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="phoneNumber"
          id="phoneNumber-label"
        >
          Phone number
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="phoneNumber"
            name="phoneNumber"
            type="tel"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="makeStyles-inputContainer-2"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-inline-5"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Password
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-describedby="password-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <p
          class="MuiFormHelperText-root"
          id="password-helper-text"
        >
          At least 6 characters
        </p>
      </div>
    </div>
    <div
      class="makeStyles-terms-3"
    >
      <div
        class="makeStyles-terms-9 makeStyles-terms-11"
      >
        <p
          class="MuiTypography-root makeStyles-info1-16 MuiTypography-body1"
        >
          <span>
            By clicking "
          </span>
          Submit
          <span>
            ", you agree to ourÂ 
          </span>
          <a
            href="https://www.care.com/terms-of-use-popup-p1030.html"
            rel="noreferrer"
            tabindex="-1"
            target="_blank"
          >
            Terms of Use
          </a>
           
          <span>
            and
          </span>
           
          <a
            href="https://www.care.com/privacy-policy-popup-p1031.html"
            rel="noreferrer"
            tabindex="-1"
            target="_blank"
          >
            Privacy Policy
          </a>
          <span>
            .
          </span>
        </p>
      </div>
    </div>
    <div
      class="makeStyles-buttonContainer-26 makeStyles-buttonContainer-27"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Submit
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
